File: Algorithm/TopCoder Algorithm Training/Go/ColorfulBoxesAndBalls.go
// ColorfulBoxesAndBalls
package main

import (
	"fmt"
)

func getMinimum(a, b int) int {
	if a >= b {
		return b
	} else {
		return a
	}
}

func getMaximum(a, b int) int {
	if a >= b {
		return a
	} else {
		return b
	}
}

type ColorfulBoxesAndBalls struct {
}

func (c ColorfulBoxesAndBalls) getMaximum(numRed, numBlue, onlyRed, onlyBlue, bothColors int) int {
	answer := -100000
	change := getMinimum(numRed, numBlue)

	for i := 0; i <= change; i++ {
		currentScore := (numRed-i)*onlyRed + (numBlue-i)*onlyBlue + 2*i*bothColors

		answer = getMaximum(answer, currentScore)
	}

	return answer
}

func main() {

	colorfulboxesandballs := ColorfulBoxesAndBalls{}
	fmt.Println(colorfulboxesandballs.getMaximum(2, 3, 100, 400, 200))
}

File: Algorithm/TopCoder Algorithm Training/C++/ColorfulBoxesAndBalls.cpp
#include <algorithm>
#include <limits.h>
#include <iostream>

using namespace std;

class ColorfulBoxesAndBalls
{
private:
	int numRed, numBlue, onlyRed, onlyBlue, bothColors;
	const int MIN_VALUE = -1000000;

public:
	int answer;
	ColorfulBoxesAndBalls(int numRed, int numBlue, int onlyRed, int onlyBlue, int bothColors);
	void getMaximum();
};


ColorfulBoxesAndBalls::ColorfulBoxesAndBalls(int numRed, int numBlue, int onlyRed, int onlyBlue, int bothColors)
{
	this->numRed = numRed;
	this->numBlue = numBlue;
	this->onlyRed = onlyRed;
	this->onlyBlue = onlyBlue;
	this->bothColors = bothColors;

	ColorfulBoxesAndBalls::getMaximum();
}

void ColorfulBoxesAndBalls::getMaximum()
{
	int answer = this->MIN_VALUE;
	int change = min(numRed, numBlue);

	for (int i = 0; i <= change; i++) 
	{
		int currentScore = (numRed - i) * onlyRed + (numBlue - i) * onlyBlue + 2 * i * bothColors;

		answer = max(answer, currentScore);
	}
	
	this->answer = answer;
}


int main()
{
	ColorfulBoxesAndBalls colorfulboxesandballs(2, 3, 100 ,400 ,200);

	cout << colorfulboxesandballs.answer << endl;
}
